app/gui/src/main/java/org/mobilitydata/gtfsvalidator/app/gui/GtfsValidatorApp.java

```diff
@@ -0,0 +1,314 @@
+      File selectedFile = chooser.getSelectedFile();
+      outputDirectoryField.setText(selectedFile.toString());
+    }
+  }
+
+  private void constructAdvancedOptionsPanel(JPanel parent) {
+    advancedOptionsPanel.setBorder(
+        BorderFactory.createCompoundBorder(
+            BorderFactory.createEmptyBorder(0, 0, 20, 0),
+            BorderFactory.createTitledBorder("Advanced Options")));
+    parent.add(advancedOptionsPanel);
+
+    JPanel panel = new JPanel();
+    panel.setLayout(new GridBagLayout());
+    advancedOptionsPanel.add(panel);
+
+    GridBagConstraints labelConstraints = new GridBagConstraints();
+    labelConstraints.gridx = 0;
+    labelConstraints.anchor = GridBagConstraints.NORTHWEST;
+    labelConstraints.ipadx = 10;
+
+    GridBagConstraints fieldConstraints = new GridBagConstraints();
+    fieldConstraints.gridx = 1;
+    fieldConstraints.anchor = GridBagConstraints.NORTHEAST;
+
+    labelConstraints.gridy = 0;
+    panel.add(new JLabel("Number of threads:"), labelConstraints);
+
+    fieldConstraints.gridy = 0;
+    panel.add(numThreadsSpinner, fieldConstraints);
+    numThreadsSpinner.setValue(1);
+
+    labelConstraints.gridy = 1;
+    panel.add(new JLabel("Country Code:"), labelConstraints);
+
+    fieldConstraints.gridy = 1;
+    panel.add(countryCodeField, fieldConstraints);
+
+    advancedOptionsPanel.setVisible(false);
+  }
+
+  private void constructValidateButton(JPanel panel) {
+    JPanel validateButtonPanel = new JPanel();
+    validateButtonPanel.setLayout(new BoxLayout(validateButtonPanel, BoxLayout.LINE_AXIS));
+    panel.add(validateButtonPanel);
+
+    JButton advancedButton = new JButton("Advanced");
+    advancedButton.addActionListener(
+        (e) -> {
+          advancedOptionsPanel.setVisible(!advancedOptionsPanel.isVisible());
+          pack();
+        });
+    validateButtonPanel.add(advancedButton);
+
+    validateButtonPanel.add(Box.createHorizontalGlue());
+
+    validateButton.setText("Validate");
+    updateValidationButtonStatus();
+    validateButton.addActionListener((e) -> runValidation());
+    validateButtonPanel.add(validateButton);
+  }
+
+  private void updateValidationButtonStatus() {
+    validateButton.setEnabled(isValidationReadyToRun());
+  }
+
+  private boolean isValidationReadyToRun() {
+    if (gtfsInputField.getText().isBlank()) {
+      return false;
+    }
+    if (outputDirectoryField.getText().isBlank()) {
+      return false;
+    }
+    return true;
+  }
+
+  private void runValidation() {
+    try {
+      ValidationRunnerConfig config = buildConfig();
+      validationRunner.run(config, this);
+    } catch (Throwable ex) {
+      validationDisplay.handleError(ex);
+    }
+  }
+
+  private ValidationRunnerConfig buildConfig() throws URISyntaxException {
+    ValidationRunnerConfig.Builder config = ValidationRunnerConfig.builder();
+    config.setPrettyJson(true);
+
+    String gtfsInput = gtfsInputField.getText();
+    if (!gtfsInput.isBlank()) {
```

Comment:
At this stage, we know that `gtfsInput` and `outputDirectory` must not be blank because the listener uses the `isValidationReadyToRun` method. Is it best practice to double check here? I guess so, I'm just curious to get your input on this. Also, it seems that the auto values assigned by the builder if those fields were blank would cause the validation to fail, right?

Reply:
In general, the further the distance between two code points that make some assumption about an input value, the more inclined I am to check it as a defensive mechanism.  Since the input validation and the config builder were far enough apart, I added the check here.

That said, you are also right that the config builder will throw an exception if the inputs aren't specified.  So we could change this code to use a Precondition check to throw an explicit exception if the inputs are blank, indicating a logic error.

Thoughts?

Reply:
I understand, it makes sense. Thank you!

I think a Precondition check to throw an explicit exception would be a good thing.

Reply:
Fixed.